python/cudf/cudf/core/dataframe.py

```diff
@@ -2770,36 +2794,48 @@ def rename(self, mapper=None, columns=None, copy=True, inplace=False):
         Notes
         -----
         Difference from pandas:
-          * Support axis='columns' only.
-          * Not supporting: index, level
+            * Not supporting: level
 
         Rename will not overwite column names. If a list with duplicates is
         passed, column names will be postfixed with a number.
         """
-        # Pandas defaults to using columns over mapper
-        if columns:
-            mapper = columns
+        if mapper is None and index is None and columns is None:
+            return self.copy(deep=copy)
 
-        out = DataFrame(index=self.index)
-        if isinstance(mapper, Mapping):
+        index = mapper if index is None and axis in (0, "index") else index
+        columns = (
+            mapper if columns is None and axis in (1, "columns") else columns
+        )
+
+        if index:
+            out = DataFrame(
+                index=list(map(lambda x: index.get(x, x), self.index))
```

Comment:
You can't run a `map` or any other iteration based operation against GPU backed objects. If we iterate in Python, then it causes it to copy from the GPU to a host object for each element which is the absolutely worst case performance scenario.

I believe we have a `replace` or `find_and_replace` API that you should be able to use to do this efficiently on the GPU.

Reply:
OK. That makes sense. I went down that route once and for some reason ended up here. My bad. I will get at it and post to slack if I get stuck.